File: tocopedia_backend/src/models/product.js
const mongoose = require("mongoose");
const { categorySchema } = require("./category");
const validator = require("validator");
const ObjectID = mongoose.Schema.Types.ObjectId;

const productSchema = new mongoose.Schema(
  {
    owner: {
      type: ObjectID,
      required: true,
      ref: "User",
    },
    name: {
      type: String,
      required: true,
      trim: true,
      maxlLength: 70,
    },
    images: {
      type: [
        {
          type: String,
          required: true,
        },
      ],
      validate: [
        (val) => val.length >= 1 && val.length <= 5,
        "product must have 1 - 5 images",
      ],
    },
    price: {
      type: Number,
      required: true,
      min: 1,
    },
    stock: {
      type: Number,
      default: 1,
      min: 0,
      validate: [
        (value) => {
          return validator.isInt(value + "");
        },
        "stock must be an integer",
      ],
    },
    SKU: {
      type: String,
      required: false,
      maxlLength: 40,
    },
    description: {
      type: String,
      required: true,
      maxlLength: 2000,
    },
    status: {
      type: String,
      enum: ["active", "inactive", "banned"],
      default: "active",
    },
    category: {
      type: ObjectID,
      required: true,
      ref: "Category",
    },
    total_sold: {
      type: Number,
      default: 0,
      validate: [
        (value) => {
          return validator.isInt(value + "");
        },
        "total_sold must be an integer",
      ],
    },
    total_rating: {
      type: Number,
      required: false,
      validate: [
        (value) => {
          return validator.isInt(value + "");
        },
        "total_rating must be an integer",
      ],
    },
    average_rating: {
      type: Number,
      required: false,
      min: 0,
      max: 5,
    },
  },
  {
    timestamps: true,
  }
);

productSchema.pre("save", async function (next) {
  const product = this;
  if (product.stock == 0) {
    product.status = "inactive";
  }

  next();
});

productSchema.methods.updateRating = async function (
  rating_count,
  rating_average
) {
  const product = this;
  product.total_rating = rating_count;
  product.average_rating = rating_average;

  await product.save();
};

const Product = mongoose.model("Product", productSchema);
module.exports = { Product, productSchema };

File: tocopedia-flutter/lib/domains/entities/product.dart
import 'package:tocopedia/domains/entities/category.dart';
import 'package:tocopedia/domains/entities/user.dart';

class Product {
  Product({
    required this.id,
    required this.owner,
    required this.name,
    required this.images,
    required this.price,
    required this.stock,
    required this.sku,
    required this.description,
    required this.status,
    required this.category,
    required this.totalSold,
    required this.totalRating,
    required this.averageRating,
    required this.createdAt,
    required this.updatedAt,
    required this.v,
  });

  final String? id;
  final User? owner;
  final String? name;
  final List<String>? images;
  final int? price;
  final int? stock;
  final String? sku;
  final String? description;
  final String? status;
  final Category? category;
  final int? totalSold;
  final int? totalRating;
  final double? averageRating;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final int? v;
}
